src/wp-includes/abilities-api/class-wp-ability.php

```diff
@@ -0,0 +1,588 @@
+	 * @see WP_Ability::check_permissions()
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param mixed $input Optional. The input data for permission checking. Default `null`.
+	 * @return bool|\WP_Error Whether the ability has the necessary permission.
+	 */
+	public function has_permission( $input = null ) {
+		_deprecated_function( __METHOD__, '6.9.0' );
+		return $this->check_permissions( $input );
+	}
+
+	/**
+	 * Executes the ability callback.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param mixed $input Optional. The input data for the ability. Default `null`.
+	 * @return mixed|\WP_Error The result of the ability execution, or WP_Error on failure.
+	 */
+	protected function do_execute( $input = null ) {
+		if ( ! is_callable( $this->execute_callback ) ) {
+			return new \WP_Error(
+				'ability_invalid_execute_callback',
+				/* translators: %s ability name. */
+				sprintf( __( 'Ability "%s" does not have a valid execute callback.' ), $this->name )
+			);
+		}
+
+		return $this->invoke_callback( $this->execute_callback, $input );
+	}
+
+	/**
+	 * Validates output data against the output schema.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param mixed $output The output data to validate.
+	 * @return true|\WP_Error Returns true if valid, or a WP_Error object if validation fails.
+	 */
+	protected function validate_output( $output ) {
+		$output_schema = $this->get_output_schema();
+		if ( empty( $output_schema ) ) {
+			return true;
+		}
+
+		$valid_output = rest_validate_value_from_schema( $output, $output_schema, 'output' );
+		if ( is_wp_error( $valid_output ) ) {
+			return new \WP_Error(
+				'ability_invalid_output',
+				sprintf(
+					/* translators: %1$s ability name, %2$s error message. */
+					__( 'Ability "%1$s" has invalid output. Reason: %2$s' ),
+					$this->name,
+					$valid_output->get_error_message()
+				)
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Executes the ability after input validation and running a permission check.
+	 * Before returning the return value, it also validates the output.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param mixed $input Optional. The input data for the ability. Default `null`.
+	 * @return mixed|\WP_Error The result of the ability execution, or WP_Error on failure.
+	 */
+	public function execute( $input = null ) {
+		$has_permissions = $this->check_permissions( $input );
+		if ( true !== $has_permissions ) {
+			if ( is_wp_error( $has_permissions ) ) {
+				if ( 'ability_invalid_input' === $has_permissions->get_error_code() ) {
+					return $has_permissions;
+				}
+				// Don't leak the permission check error to someone without the correct perms.
+				_doing_it_wrong(
+					__METHOD__,
+					esc_html( $has_permissions->get_error_message() ),
+					'6.9.0'
+				);
+			}
+
+			return new \WP_Error(
+				'ability_invalid_permissions',
+				/* translators: %s ability name. */
+				sprintf( __( 'Ability "%s" does not have necessary permission.' ), $this->name )
+			);
+		}
+
+		/**
+		 * Fires before an ability gets executed.
```

Comment:
```suggestion
		 * Fires before an ability gets executed and after permission check.
```

Make it clear that the permission check has already been completed. This will help ensure that in the future if there is refactoring this isn't moved to before the permission check and potentially introduce issues for extenders.